File: src/com/train/algorithm/tree/implementInC++/PathSum.cpp
//
// Created by 罗斯 on 30/10/2022.
//
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (!root) return false;
        if (!root->left && !root->right) return root->val == targetSum;
        int sum = targetSum - root->val;
        return hasPathSum(root->left, sum) || hasPathSum(root->right, sum);
    }
};

File: src/com/train/algorithm/tree/implementInPython/PathSum.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def hasPathSum(self, root, targetSum):
        if not root: return False
        if not root.left and not root.right: return root.val == targetSum
        l_sum = self.hasPathSum(root.left,targetSum - root.val)
        r_sum = self.hasPathSum(root.right, targetSum - root.val)
        return  l_sum or r_sum